pkg/skaffold/actions/runner.go

```diff
@@ -0,0 +1,167 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package actions
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"time"
+
+	"github.com/fatih/semgroup"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/GoogleContainerTools/skaffold/v2/pkg/skaffold/graph"
+	"github.com/GoogleContainerTools/skaffold/v2/pkg/skaffold/output/log"
+)
+
+type Runner struct {
+	execEnvByAction map[string]ExecEnv
+	orderedExecEnvs []ExecEnv
+	acsByExecEnv    map[ExecEnv][]string
+}
+
+func NewRunner(execEnvByAction map[string]ExecEnv, orderedExecEnvs []ExecEnv, acsByExecEnv map[ExecEnv][]string) Runner {
+	return Runner{execEnvByAction, orderedExecEnvs, acsByExecEnv}
+}
+
+func (r Runner) ExecAll(ctx context.Context, out io.Writer, allbuilds []graph.Artifact) error {
+	acs, err := r.prepareActions(ctx, out, allbuilds)
+	if err != nil {
+		return err
+	}
+
+	defer r.cleanup(ctx, out, acs, r.orderedExecEnvs)
+	return r.execWithFailingSafe(ctx, out, acs)
+}
+
+func (r Runner) Exec(ctx context.Context, out io.Writer, allbuilds []graph.Artifact, aName string) error {
+	execEnv, found := r.execEnvByAction[aName]
+	if !found {
+		return fmt.Errorf("custom action %v not found", aName)
+	}
+
+	acs, err := execEnv.PrepareActions(ctx, out, allbuilds, []string{aName})
+	if err != nil {
+		return err
+	}
+
+	// We expect only one action to be created.
+	if len(acs) != 1 {
+		return fmt.Errorf("failed to create %v action", aName)
+	}
+
+	a := acs[0]
+	a.SetExecFunc(r.getExecFunc(a))
+
+	defer r.cleanup(ctx, out, []Task{a}, []ExecEnv{execEnv})
+	return a.Exec(ctx, out)
+}
+
+func (r Runner) prepareActions(ctx context.Context, out io.Writer, allbuilds []graph.Artifact) ([]Task, error) {
+	preparedAcs := []Task{}
+	for _, execEnv := range r.orderedExecEnvs {
+		acsNames := r.acsByExecEnv[execEnv]
+		acs, err := execEnv.PrepareActions(ctx, out, allbuilds, acsNames)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, a := range acs {
+			a.SetExecFunc(r.getExecFunc(a))
+			preparedAcs = append(preparedAcs, a)
+		}
+	}
+
+	return preparedAcs, nil
+}
+
+func (r Runner) getExecFunc(a Action) ExecStrategy {
+	if a.IsFailFast() {
+		return r.execWithFailingFast
+	}
+	return r.execWithFailingSafe
+}
+
+func (r Runner) execWithFailingFast(ctx context.Context, out io.Writer, ts []Task) error {
+	g, gCtx := errgroup.WithContext(ctx)
+
+	for _, t := range ts {
+		t := t
+		g.Go(func() error {
+			return r.execute(gCtx, t, out)
+		})
+	}
+
+	return g.Wait()
+}
+
+func (r Runner) execWithFailingSafe(ctx context.Context, out io.Writer, ts []Task) error {
+	const maxWorkers = math.MaxInt64
+	g := semgroup.NewGroup(context.Background(), maxWorkers)
+
+	for _, t := range ts {
+		t := t
+		g.Go(func() error {
+			return r.execute(ctx, t, out)
+		})
+	}
+
+	return g.Wait()
+}
+
+func (r Runner) execute(ctx context.Context, t Task, out io.Writer) error {
+	var err error
+	execCh := make(chan error)
+
+	go func() {
+		execCh <- t.Exec(ctx, out)
```

Comment:
since `Task` defines a `Timeout` shouldn't it be the responsibility of the `Task` implementation to ensure that it gets cancelled on timeout. In which case, you should be able to just call `t.Exec` directly without the separate goroutine wrapping and the `select` statements since this method is essentially synchronous. WDYT?

Reply:
You are right. From the config point of view, the timeout is defined at the action level. With the latest changes I created a new `Action` struct type that handles its tasks timeout. This way, the logic to interrupt the execution due to a time out is always the same regardless if we are executing docker or k8s tasks. Let me know what do you think!